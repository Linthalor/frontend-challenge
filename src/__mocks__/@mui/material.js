const genMockComponent = (name) => (props) => {
  const MockName = `mocked-mui-${name}`;
  const {sx, ...remainingProps} = props;
  return <MockName {...remainingProps} sx={sx ? JSON.stringify(sx) : undefined} />;
}

module.exports = {
  ...jest.requireActual('@mui/material'),
  Box: genMockComponent('box'),
  Button: genMockComponent('button'),
  IconButton: genMockComponent('icon-button'),
  InputAdornment: genMockComponent('input-adornment'),
  TextField: genMockComponent('text-field'),
  Typography: genMockComponent('typography'),
  Card: genMockComponent('card'),
  CardContent: genMockComponent('card-content'),
  CardActions: genMockComponent('card-actions'),
};