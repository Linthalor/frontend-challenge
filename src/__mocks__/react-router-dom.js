module.exports = {
  ...jest.requireActual('react-router-dom'),
  Navigate: (props) => {
    const MockName = `mocked-react-router-dom-navigate`;
    return <MockName {...props} />;
  }
};