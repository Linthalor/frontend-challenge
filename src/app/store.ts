import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { signUpReducer } from '../features/sign-up/sign-up.slice';
import { colorReducer } from '../features/color/color.slice';

export const store = configureStore({
  reducer: {
    signUp: signUpReducer,
    color: colorReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
