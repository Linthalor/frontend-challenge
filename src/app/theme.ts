import { createTheme } from '@mui/material';

// Themed using Upgrade's palette
export const theme = createTheme({
  palette: {
    primary: {
      light: '#60c7a1',
      main: '#00a669',
      dark: '#008051',
    },
    secondary: {
      main: '#3f3d56',
      light: '#6bc1fe',
    },
    success: {
      main: '#03b57a',
      contrastText: '#ffffff',
    },
    error: {
      main: '#c83c4c',
      contrastText: '#ffffff',
    },
    text: {
      primary: '#3f465e',
      secondary: '#008051'
    }
  },
});