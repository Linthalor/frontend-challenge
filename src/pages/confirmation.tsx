import {
  Box,
  Button,
  Checkbox,
  CircularProgress,
  FormControlLabel,
  TextField,
  Typography,
} from '@mui/material';
import { Link, useNavigate } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { useAppDispatch } from '../app/hooks';
import { SignUpCard } from '../components/signup-card';
import {
  selectSignUpStatus,
  selectUserFormValues,
  selectUserPrefsValues,
  signUpAsync,
} from '../features/sign-up/sign-up.slice';
import { useSelector } from 'react-redux';
import { useEffect, useState } from 'react';
import { isError, isValue } from '../model/util/async-prop';
import { Error } from '@mui/icons-material';

export const Confirmation = () => {
  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const { t } = useTranslation();
  const signUpUserData = useSelector(selectUserFormValues);
  const signUpPrefsData = useSelector(selectUserPrefsValues);
  const signUpStatus = useSelector(selectSignUpStatus);
  const [submitted, setSubmitted] = useState(false);

  const onSubmit = () => {
    if (signUpUserData && signUpPrefsData) {
      dispatch(signUpAsync({
        ...signUpUserData,
        ...signUpPrefsData,
      }))
      setSubmitted(true);
    }
  };

  useEffect(() => {
    if (submitted && isValue(signUpStatus)) return navigate('/success');
    if (submitted && isError(signUpStatus)) return navigate('/error');
  }, [signUpStatus, submitted, navigate]);

  if (submitted) {
    return (
      <Box
        sx={{
          width: '100%',
          height: '100%',
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center'
          }}
        >
        <CircularProgress size={200} />
      </Box>
    );
  }

  return (
    <SignUpCard
      content={<>
        <Typography
          sx={{
            textAlign: 'center',
            fontWeight: 'bold'
          }}
          variant="h5"
          color="text.primary"
          gutterBottom
        >
          {t('page:confirmation:label')}
        </Typography>
        <TextField
          id="first-name-confirmation-field"
          label={t('form:name:label')}
          variant="outlined"
          type="text"
          value={signUpUserData?.name}
          error={!signUpUserData?.name}
          disabled
          aria-readonly
        />
        <TextField
          id="email-confirmation-field"
          label={t('form:email:label')}
          variant="outlined"
          type="email"
          value={signUpUserData?.email}
          error={!signUpUserData?.email}
          disabled
          aria-readonly
        />
        {/*
          seems like showing the censored password to the user is a little useless
          and polutes the view. Omitting it.
        */}
        {/* <TextField
          id="password-confirmation-field"
          label={t('form:password:label')}
          variant="outlined"
          type="password"
          value={signUpUserData?.password}
          disabled
          aria-readonly
        /> */}
        <TextField
          id="favorite-color-confirmation-field"
          label={t('form:color:label')}
          variant="outlined"
          type="text"
          value={signUpPrefsData?.color}
          error={!signUpPrefsData?.color}
          disabled
          aria-readonly
        />
        <Box sx={{ display: 'flex', flexDirection: 'row', alignItems: 'center' }}>
          {!signUpPrefsData?.terms && <Error color="error" sx={{ mr: 1}} />}
          <FormControlLabel
            label={
              <Box
                sx={{
                  display: 'flex',
                  flexDirection: 'row',
                  alignItems: 'center'
                }}
              >
                <Typography>
                  {t('common:terms:agree')}
                  {' '}
                  {t('common:terms:termsAndConditions')}
                </Typography>
              </Box>
            }
            control={
              <Checkbox
                disabled
                checked={signUpPrefsData?.terms}
              />
            }
          />
        </Box>
      </>}
      actions={<>
        <Button variant='outlined' color='secondary' component={Link} to="/more-info">
          {t('common:back')}
        </Button>
        <Button
          variant="contained"
          type="submit"
          disabled={!signUpUserData || !signUpPrefsData}
          onClick={onSubmit}
        >
          {t('common:submit')}
        </Button>
      </>}
    />
  );
}