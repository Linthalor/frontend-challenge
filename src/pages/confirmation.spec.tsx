import { cleanup, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createTestStore, store as realStore } from '../app/store';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from '@mui/material';
import { theme } from '../app/theme';
import { Confirmation } from './confirmation';

jest.mock('@mui/material');

describe('Confirmation', () => {
  let store: typeof realStore;
  beforeEach(() => {
    store = createTestStore();
  });
  afterEach(() => {
    cleanup();
  });

  it('renders as expected', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <BrowserRouter>
            <Confirmation />
          </BrowserRouter>
        </Provider>
      </ThemeProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});