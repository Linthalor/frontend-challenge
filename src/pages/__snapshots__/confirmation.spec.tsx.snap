// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirmation renders as expected 1`] = `
<mocked-mui-box
  sx="{\\"width\\":\\"100%\\",\\"height\\":\\"100%\\",\\"display\\":\\"flex\\",\\"justifyContent\\":\\"center\\",\\"alignItems\\":\\"center\\"}"
>
  <mocked-mui-card
    sx="{\\"width\\":400,\\"height\\":\\"unset\\",\\"display\\":\\"flex\\",\\"flexDirection\\":\\"column\\",\\"justifyContent\\":\\"center\\"}"
  >
    <mocked-mui-card-content
      sx="{\\"display\\":\\"flex\\",\\"flexDirection\\":\\"column\\",\\"alignItems\\":\\"stretch\\",\\"justifyContent\\":\\"start\\",\\"gap\\":1,\\"flex\\":1}"
    >
      <mocked-mui-typography
        color="text.primary"
        gutterbottom="true"
        sx="{\\"textAlign\\":\\"center\\",\\"fontWeight\\":\\"bold\\"}"
        variant="h5"
      >
        Confirmation
      </mocked-mui-typography>
      <mocked-mui-text-field
        aria-readonly="true"
        disabled="true"
        error="true"
        id="first-name-confirmation-field"
        label="First Name"
        type="text"
        variant="outlined"
      />
      <mocked-mui-text-field
        aria-readonly="true"
        disabled="true"
        error="true"
        id="email-confirmation-field"
        label="Email"
        type="email"
        variant="outlined"
      />
      <mocked-mui-text-field
        aria-readonly="true"
        disabled="true"
        error="true"
        id="favorite-color-confirmation-field"
        label="Favorite Color"
        type="text"
        variant="outlined"
      />
      <mocked-mui-box
        sx="{\\"display\\":\\"flex\\",\\"flexDirection\\":\\"row\\",\\"alignItems\\":\\"center\\"}"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-1esn0is-MuiSvgIcon-root"
          data-testid="ErrorIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
          />
        </svg>
        <label
          class="MuiFormControlLabel-root Mui-disabled MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            aria-disabled="true"
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-disabled css-1uxrgnw-MuiButtonBase-root-MuiCheckbox-root"
            tabindex="-1"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              disabled=""
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label Mui-disabled css-ahj2mt-MuiTypography-root"
          >
            <mocked-mui-box
              sx="{\\"display\\":\\"flex\\",\\"flexDirection\\":\\"row\\",\\"alignItems\\":\\"center\\"}"
            >
              <mocked-mui-typography>
                I Agree to the
                 
                Terms & Conditions
              </mocked-mui-typography>
            </mocked-mui-box>
          </span>
        </label>
      </mocked-mui-box>
    </mocked-mui-card-content>
    <mocked-mui-card-actions
      sx="{\\"display\\":\\"flex\\",\\"flexDirection\\":\\"row\\",\\"justifyContent\\":\\"end\\",\\"gap\\":1}"
    >
      <mocked-mui-button
        color="secondary"
        component="[object Object]"
        to="/more-info"
        variant="outlined"
      >
        Back
      </mocked-mui-button>
      <mocked-mui-button
        disabled="true"
        type="submit"
        variant="contained"
      >
        Submit
      </mocked-mui-button>
    </mocked-mui-card-actions>
  </mocked-mui-card>
</mocked-mui-box>
`;
